src/net/server/dgram.rs

```diff
@@ -0,0 +1,745 @@
+//! [Datagram]: https://en.wikipedia.org/wiki/Datagram
+use core::fmt::Debug;
+use core::future::poll_fn;
+use core::ops::Deref;
+use core::sync::atomic::Ordering;
+use core::time::Duration;
+
+use std::io;
+use std::net::SocketAddr;
+use std::string::String;
+use std::string::ToString;
+use std::sync::{Arc, Mutex};
+
+use octseq::Octets;
+use tokio::io::ReadBuf;
+use tokio::net::UdpSocket;
+use tokio::sync::watch;
+use tokio::time::interval;
+use tokio::time::timeout;
+use tokio::time::Instant;
+use tokio::time::MissedTickBehavior;
+use tracing::Level;
+use tracing::{enabled, error, trace, warn};
+
+use crate::base::Message;
+use crate::net::server::buf::BufSource;
+use crate::net::server::error::Error;
+use crate::net::server::message::MessageProcessor;
+use crate::net::server::message::{MessageDetails, Request};
+use crate::net::server::metrics::ServerMetrics;
+use crate::net::server::middleware::chain::MiddlewareChain;
+use crate::net::server::service::{CallResult, Service, ServiceFeedback};
+use crate::net::server::sock::AsyncDgramSock;
+use crate::net::server::util::to_pcap_text;
+use crate::utils::config::DefMinMax;
+
+use super::buf::VecBufSource;
+use super::message::{TransportSpecificContext, UdpSpecificTransportContext};
+use super::middleware::builder::MiddlewareBuilder;
+use super::service::ServerCommand;
+use crate::base::wire::Composer;
+
+/// A UDP transport based DNS server transport.
+///
+/// UDP aka User Datagram Protocol, as implied by the name, is a datagram
+/// based protocol. This type defines a type of [`DgramServer`] that expects
+/// connections to be received via [`tokio::net::UdpSocket`] and can thus be
+/// used to implement a UDP based DNS server.
+pub type UdpServer<Svc> = DgramServer<UdpSocket, VecBufSource, Svc>;
+
+/// Limit the time to wait for a complete message to be written to the client.
+///
+/// The value has to be between 1ms and 60 seconds. The default value is 5
+/// seconds.
+const WRITE_TIMEOUT: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(5),
+    Duration::from_millis(1),
+    Duration::from_secs(60),
+);
+
+/// Limit suggested for the maximum response size to create.
+///
+/// The value has to be between 512 and 4,096 per [RFC 6891]. The default
+/// value is 1232 per the [2020 DNS Flag Day].
+///
+/// The [`Service`] and [`MiddlewareChain`] (if any) are response for
+/// enforcing this limit.
+///
+/// [2020 DNS Flag Day]: http://www.dnsflagday.net/2020/
+/// [RFC 6891]: https://datatracker.ietf.org/doc/html/rfc6891#section-6.2.5
+const MAX_RESPONSE_SIZE: DefMinMax<u16> = DefMinMax::new(1232, 512, 4096);
+
+//----------- Config ---------------------------------------------------------
+
+/// Configuration for a datagram server.
+#[derive(Clone, Debug)]
+pub struct Config<RequestOctets, Target>
+where
+    RequestOctets: Octets,
+    Target: Composer + Default,
+{
+    /// Limit suggested to [`Service`] on maximum response size to create.
+    max_response_size: Option<u16>,
```

Comment:
So, the Unbound man page says:
```
max-udp-size: <number>
              Maximum UDP response size (not applied to TCP response).   65536
              disables the udp response size maximum, and uses the choice from
              the client, always.  Suggested values are 512 to  4096.  Default
              is  1232.  The  default  value  is  the  same as the default for
              edns-buffer-size.
```

That seems to say "if disabled use the client value" which is what the code in this PR does: if a library consumer calls `dgram::Config::set_max_response_size(None)` (to disable the default 1232 value) then the `EdnsMiddlewareProcessor` will use whichever is larger: the EDNS requestors udp payload size (taken from the RFC 6891 OPT record `CLASS` field described as "requestor's UDP payload size"), or 512, i.e. "uses the choice from the client, always".

You said:
> I think the last part is wrong. max_response_size_hint has to be the minimum of max_response_size and what is in the EDNS(0) OPT.

Which seems to contradict what the Unbound man page says and indeed what the code in this PR does.

You also said:
> If there is no EDNS(0) OPT then it should be 512.

The code in this PR says if there is no EDNS(0) OPT record from the client then use the server configured value which defaults to the flag day value 1232, or can be set to something else in the range 512 <= n <= 4096, or disabled entirely with `None` which in combination with no EDNS(0) OPT record from the client would result in a `None` value for `max_response_size_hint` being propagated (a) to the `Service` impl, to do with as it sees fit, and then (b) back down the middleware chain which, if it includes the `EdnsMiddlewareProcessor` (included by default), will NOT truncate because the hint is `None`.

So, I'm confused ;-) Please advise.

Reply:
Let's start simple. The reply that is sent to the user (over UDP) can never be larger than what is specified in the EDNS(0) OPT record (or 512 if the value is lower than that), or if the OPT record is not there, 512. That is what the RFCs say. Maybe you can adjust your interpretation of what the unbound manual says to this requirement. No matter what the unbound manual says, we are not going to violate RFCs.

Reply:
Hi @Philip-NLnetLabs,

Please see if commit 18c3640ce987603fcbec33fab1e1750c5a705366 is an improvement.

Cheers,

Reply:
Yes, that looks good